[PyTorch Edge] Conditionally trim dispatch key set to save heap memory at runtime

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65732

For certain on-device uses, runtime memory comes at a premium. On-device deployments won't use all the available dispatch keys, so it makes sense to keep only the on-device specific ones around for such uses to reduce runtime heap memory allocated.

This change keeps just 10 dispatch keys (the ones that used on-device), guarded under the `C10_MOBILE_TRIM_DISPATCH_KEYS` macro. it tries to keep the other code-paths unaffected and uses `constexpr` for use in the `array` declaration, and simple inline functions to ensure that the compiler is able to optimize these for server builds.

Differential Revision: [D31185407](https://our.internmc.facebook.com/intern/diff/D31185407/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31185407/)!

dhruvbird committed:
[PyTorch Edge] Conditionally trim dispatch key set to save heap memory at runtime

For certain on-device uses, runtime memory comes at a premium. On-device deployments won't use all the available dispatch keys, so it makes sense to keep only the on-device specific ones around for such uses to reduce runtime heap memory allocated.

This change keeps just 10 dispatch keys (the ones that used on-device), guarded under the `C10_MOBILE_TRIM_DISPATCH_KEYS` macro. it tries to keep the other code-paths unaffected and uses `constexpr` for use in the `array` declaration, and simple inline functions to ensure that the compiler is able to optimize these for server builds.

Differential Revision: [D31185407](https://our.internmc.facebook.com/intern/diff/D31185407/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31185407/)!

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -299,8 +299,11 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
-  auto dispatch_ix = static_cast<uint8_t>(dispatch_key);
-  dispatchTable_[dispatch_ix] = computeDispatchTableEntry(dispatcher, dispatch_key);
+  auto dispatch_ix = c10::getMappedDispatchKey(dispatch_key);
+  if (dispatch_ix == -1) {
```
I'd like this condition to inline away on server and I don't see any indication that it will

swolchok replied:
clang seems to get this right. as of C++17, it appears to be UB to get an out-of-range value into an enum, so the compiler can prove that the return value of getMappedDispatchKey is never -1. Also, the change in C++17 was a defect in the standard, so you get the correct behavior in C++14 mode as well if hte compiler implements the fix. See https://godbolt.org/z/6f194f93M and https://stackoverflow.com/questions/18195312/what-happens-if-you-static-cast-invalid-value-to-enum-class

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -170,7 +170,11 @@ class TORCH_API OperatorEntry final {
   const KernelFunction& lookup(DispatchKey k) const {
-    const auto& kernel = dispatchTable_[static_cast<uint8_t>(k)];
+    auto idx = getMappedDispatchKey(k);
+    if (idx == -1) {
```
Ditto here (and for mobile, you probably want to C10_UNLIKELY this)

ezyang reviewed (approved):
TeStInG sTrAtEgY

dhruvbird committed:
Update on "[PyTorch Edge] Conditionally trim dispatch key set to save heap memory at runtime"


For certain on-device uses, runtime memory comes at a premium. On-device deployments won't use all the available dispatch keys, so it makes sense to keep only the on-device specific ones around for such uses to reduce runtime heap memory allocated.

This change keeps just 10 dispatch keys (the ones that used on-device), guarded under the `C10_MOBILE_TRIM_DISPATCH_KEYS` macro. it tries to keep the other code-paths unaffected and uses `constexpr` for use in the `array` declaration, and simple inline functions to ensure that the compiler is able to optimize these for server builds.

Differential Revision: [D31185407](https://our.internmc.facebook.com/intern/diff/D31185407/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31185407/)!

[ghstack-poisoned]

dhruvbird commented:
> What's the test strategy. In particular, how can I emulate this mode on server (for easy debugging), and is this adequately covered by OSS tests.

@ezyang I've added `PYTORCH_MOBILE_TRIM_DISPATCH_KEY_SET` to allow PT to be built under this mode.

dhruvbird committed:
Update on "[PyTorch Edge] Conditionally trim dispatch key set to save heap memory at runtime"


For certain on-device uses, runtime memory comes at a premium. On-device deployments won't use all the available dispatch keys, so it makes sense to keep only the on-device specific ones around for such uses to reduce runtime heap memory allocated.

This change keeps just 10 dispatch keys (the ones that used on-device), guarded under the `C10_MOBILE_TRIM_DISPATCH_KEYS` macro. it tries to keep the other code-paths unaffected and uses `constexpr` for use in the `array` declaration, and simple inline functions to ensure that the compiler is able to optimize these for server builds.

Differential Revision: [D31185407](https://our.internmc.facebook.com/intern/diff/D31185407/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31185407/)!

[ghstack-poisoned]

dhruvbird commented:
Updated build/test plan on internal FB diff.

dhruvbird committed:
Update on "[PyTorch Edge] Conditionally trim dispatch key set to save heap memory at runtime"


For certain on-device uses, runtime memory comes at a premium. On-device deployments won't use all the available dispatch keys, so it makes sense to keep only the on-device specific ones around for such uses to reduce runtime heap memory allocated.

This change keeps just 10 dispatch keys (the ones that used on-device), guarded under the `C10_MOBILE_TRIM_DISPATCH_KEYS` macro. it tries to keep the other code-paths unaffected and uses `constexpr` for use in the `array` declaration, and simple inline functions to ensure that the compiler is able to optimize these for server builds.

Differential Revision: [D31185407](https://our.internmc.facebook.com/intern/diff/D31185407/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31185407/)!

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Edge] Conditionally trim dispatch key set to save heap memory at runtime"


For certain on-device uses, runtime memory comes at a premium. On-device deployments won't use all the available dispatch keys, so it makes sense to keep only the on-device specific ones around for such uses to reduce runtime heap memory allocated.

This change keeps just 10 dispatch keys (the ones that used on-device), guarded under the `C10_MOBILE_TRIM_DISPATCH_KEYS` macro. it tries to keep the other code-paths unaffected and uses `constexpr` for use in the `array` declaration, and simple inline functions to ensure that the compiler is able to optimize these for server builds.

Differential Revision: [D31185407](https://our.internmc.facebook.com/intern/diff/D31185407/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31185407/)!

[ghstack-poisoned]
